pkg/gadget-collection/gadgets/top/ebpf/gadget.go

```diff
@@ -0,0 +1,209 @@
+	t := `ebpftop shows cpu time used by ebpf programs.
+
+The following parameters are supported:
+ - %s: Output interval, in seconds. (default %d)
+ - %s: Maximum rows to print. (default %d)
+ - %s: The field to sort the results by (%s). (default %s)`
+	return fmt.Sprintf(t, types.IntervalParam, types.IntervalDefault,
+		types.MaxRowsParam, types.MaxRowsDefault,
+		types.SortByParam, strings.Join(types.SortBySlice, ","), types.SortByDefault)
+}
+
+func (f *TraceFactory) OutputModesSupported() map[string]struct{} {
+	return map[string]struct{}{
+		"Stream": {},
+	}
+}
+
+func deleteTrace(name string, t interface{}) {
+	trace := t.(*Trace)
+	if trace.started {
+		trace.stop()
+	}
+}
+
+func (f *TraceFactory) Operations() map[string]gadgets.TraceOperation {
+	n := func() interface{} {
+		return &Trace{
+			helpers: f.Helpers,
+		}
+	}
+
+	return map[string]gadgets.TraceOperation{
+		"start": {
+			Doc: "Start ebpftop gadget",
+			Operation: func(name string, trace *gadgetv1alpha1.Trace) {
+				f.LookupOrCreate(name, n).(*Trace).Start(trace)
+			},
+		},
+		"stop": {
+			Doc: "Stop ebpftop gadget",
+			Operation: func(name string, trace *gadgetv1alpha1.Trace) {
+				f.LookupOrCreate(name, n).(*Trace).Stop(trace)
+			},
+		},
+	}
+}
+
+func (t *Trace) Start(trace *gadgetv1alpha1.Trace) {
+	if t.started {
+		trace.Status.State = "Started"
+		return
+	}
+
+	t.traceName = gadgets.TraceName(trace.ObjectMeta.Namespace, trace.ObjectMeta.Name)
+	t.node = trace.Spec.Node
+
+	maxRows := types.MaxRowsDefault
+	intervalSeconds := types.IntervalDefault
+	sortBy := types.SortByDefault
+
+	if trace.Spec.Parameters != nil {
+		params := trace.Spec.Parameters
+		var err error
+
+		if val, ok := params[types.MaxRowsParam]; ok {
+			maxRows, err = strconv.Atoi(val)
+			if err != nil {
+				trace.Status.OperationError = fmt.Sprintf("%q is not valid for %s: %v", val, types.MaxRowsParam, err)
+				return
+			}
+		}
+
+		if val, ok := params[types.IntervalParam]; ok {
+			intervalSeconds, err = strconv.Atoi(val)
+			if err != nil {
+				trace.Status.OperationError = fmt.Sprintf("%q is not valid for %s: %v", val, types.IntervalParam, err)
+				return
+			}
+		}
+
+		if val, ok := params[types.SortByParam]; ok {
+			sortBy, err = types.ParseSortBy(val)
+			if err != nil {
+				trace.Status.OperationError = fmt.Sprintf("%q is not valid for %s: %v", val, types.SortByParam, err)
+				return
+			}
+		}
+	}
+
+	config := &ebpftoptracer.Config{
+		MaxRows:  maxRows,
+		Interval: time.Second * time.Duration(intervalSeconds),
+		SortBy:   sortBy,
+	}
+
+	statsCallback := func(stats []types.Stats) {
+		ev := types.Event{
+			Stats: stats,
+		}
+		ev.Node = trace.Spec.Node
+
+		r, err := json.Marshal(ev)
+		if err != nil {
+			log.Warnf("Gadget %s: Failed to marshal event: %s", trace.Spec.Gadget, err)
+			return
+		}
+		t.helpers.PublishEvent(t.traceName, string(r))
+	}
+
+	errorCallback := func(err error) {
+		ev := types.Event{
+			Error: fmt.Sprintf("Gadget failed with: %v", err),
```

Comment:
You can directly use `eventtypes.Err(fmt.Sprintf("Gadget failed with: %v", err))` . -> https://github.com/kinvolk/inspektor-gadget/blob/v0.7.0/pkg/types/types.go#L77

Reply:
This is the same way it's used in the other gadgets (copied it from there -> https://github.com/kinvolk/inspektor-gadget/blob/v0.7.0/pkg/gadget-collection/gadgets/top/block-io/gadget.go#L167). Maybe it makes more sense to change it later everywhere in one go.